python/python/bystro/covariance/_covariance_np.py

```diff
@@ -0,0 +1,96 @@
+"""
+This implements two objects, the empirical covariance estimator used as a 
+baseline comparison method. It also implements BayesianCovariance, which
+implements MAP estimation using several common priors.
+
+Objects
+-------
+EmpiricalCovariance(BaseCovariance)
+    
+
+"""
+import numpy as np
+from ._base_covariance import BaseCovariance
+from copy import deepcopy
+
+
+class EmpiricalCovariance(BaseCovariance):
+    def __init__(self):
+        """
+        This object just fits the covariance matrix as the standard sample
+        covariance matrix
+        """
+        super().__init__()
+
+    def fit(self, X):
+        """
+        This fits a covariance matrix using samples X.
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,n_covariates)
+            The data
+        """
+        self.N, self.p = X.shape
+        XTX = np.dot(X.T, X)
+        self.covariance = XTX / self.N
```

Comment:
Would np.cov not work as a drop-in replacement here?  (And do we want to divide by N or N-1 here?)  Does this implementation work for non-centered matrices?  

I think I'm just confused about what X is and what properties we can expect it to have, could we gloss that at the beginning of the module

Reply:
Yeah I think so. The reason for X is it allows for some stochasticity, which I like in tests like this.

Reply:
Sorry, could you unpack that a bit more? I'm not sure I'm following.  My questions (consolidating this thread and the other one on the same topic) are:

1.  This implementation only computes the covariance matrix if X is already centered, i.e. E[X_i] = E[X_j] 0 for all rows i and columns j.  Are we guaranteed that X will always be centered?  If so, can we indicate that somehow in the code, like with a pre-condition check or even a descriptive variable name?  If X is not guaranteed to be centered but this is the implementation we want, should we call it something like `non_centered_covariance_matrix` instead?  Maybe it's a terminological confusion between fields but I've always heard of covariance matrices referring to the centered version, and if the opposite convention holds here it seems noteworthy enough to document.

2.  If we do want the centered version, why not just use `np.cov`?  I'm not saying that we should use libraries for everything, but numpy seems like a safe assumption as a dependency for an ML / stats module.  If we're doing something by hand when there's a very natural approach with a library function, it raises a question in the reader's mind why we're doing something by hand instead and what the difference might be between our approach and the library function.  In this case, there is in fact a difference (the non-centered vs. centered issue) and if that difference is intentional I think we want to signpost it to the reader.
3.  We're dividing by N rather than (N-1), meaning that our estimates of the covariance will be biased (though not by much if N is large).  I'm not even worried so much about the bias as about the difficulty the next engineer might have in interpreting the intent behind the choice of biased estimator.

Are these questions roughly checking out?  I'm not as familiar with covariance matrix estimation so I might just be misunderstanding the intent here, but I think these are questions the modal engineer might also have when reading this code.

Reply:
np.std uses the biased estimator as a default, I assumed that np.cov did too but apparently I was wrong. Either way, by writing stuff out there's no ambiguity. I'm making it clear that we should expect centered data. Should I make an assert statement? Also for (3), since we are using maximum likelihood for the remaining approaches might as well stay consistent. The biased version is the ML version.

Also centering is only applicable in the columns.

Reply:
I think where I land here is that if we have a class called EmpiricalCovariance, that class needs to compute empirical covariances.

Since empirical covariance is a well-defined mathematical concept that is unlikely to change over the lifetime of the project, my recommendation (as I understand the intent of your code) would be use `np.cov` or to implement the more or less one-line formula as it appears in stats textbooks, so that callers of this class don't get surprised.

If this class only computes empirical covariances for centered matrices and not for others, and this is correct and necessary for your purposes, then:

1.  The restricted domain should be reflected in the class name, because it's no longer computing the empirical covariance as every textbook defines it, but some other quantity.  Maybe we could call it `CenteredEmpiricalCovariance` or something.
2. As per a previous comment, if we're writing things out that are almost but not exactly equivalent to common library functions, that actually introduces ambiguity because it invites the reader to wonder, as I did, why we're not either using a common library functions or fully implementing the well-known formula for empirical covariance ourselves.  But given all that we've written back and forth here trying to understand each other, I think it's at least worth a comment in the code.  The next engineer will not even have the benefit of this record of correspondence.
3. If the `fit` method of the `EmpiricalCovariance` class receives non-centered input, it will not compute the empirical covariance of that input, but silently return some other output instead.  I would recommend it just refuse to calculate anything in those cases.  The conventional way to do that is to raise a ValueError.  If we check for value errors with asserts, we risk silent failures as discussed in another thread on this PR.

I want to leave open the possibility that I've just entirely misunderstood the intent of this code, in which case I apologize, but please consider this as a good faith attempt to understand the intent here and consider that future engineers may have similar questions.

Thank you for noting the typo, corrected above.